# 1장 2강: 유의수준·검정력·표본 크기의 트레이드오프와 실질적 유의성 — 실습문제

## 실습 목표

- 효과 크기와 유의수준이 같을 때 표본 크기에 따라 검정력이 어떻게 변하는지 설명할 수 있다.
- 목표 검정력과 유의수준을 이용해 필요한 표본 크기를 계산할 수 있다.
- 두 집단의 평균 차이에 대한 Cohen’s d를 계산하고 해석할 수 있다.
- p-value와 효과 크기, 실무 기준을 함께 사용하여 의사결정 근거를 작성할 수 있다.

## 실습 환경 / 데이터

- Python
- NumPy, pandas
- scipy.stats
- statsmodels
- `ames_housing.csv`

주요 컬럼은 다음과 같습니다.

| 컬럼 | 의미 |
|---|---|
| `SalePrice` | 주택 판매가격 |
| `CentralAir` | 중앙 냉방시설 유무(`Y`, `N`) |
| `KitchenQual` | 주방 품질(`Ex`, `Gd`, `TA`, `Fa`) |

> 모든 검정은 별도 지시가 없으면 유의수준 `α = 0.05`를 사용합니다.  
> Cohen’s d는 절댓값을 기준으로 약 0.2는 작은 효과, 0.5는 중간 효과, 0.8 이상은 큰 효과로 해석합니다.

## 실습 준비

1. 필요한 라이브러리를 불러오세요.
2. Ames Housing 데이터를 `df`에 불러오세요.
3. 데이터 크기, 전체 결측치 수, 컬럼명과 상위 5개 행을 확인하세요.

In [1]:
# 실습 준비 코드를 작성하세요.

import numpy as np
import pandas as pd
from scipy import stats

df = pd.read_csv('ames_housing.csv', encoding='utf-8-sig')

print('데이터 크기 :', df.shape)
print('전체 결측치 수 :', df.isnull().sum().sum())
df.head()

데이터 크기 : (1460, 10)
전체 결측치 수 : 0


,SalePrice,GrLivArea,LotArea,OverallQual,KitchenQual,CentralAir,HeatingQC,PavedDrive,Neighborhood,YearBuilt
0,208500,1710,8450,7,Gd,Y,Ex,Y,CollgCr,2003
1,181500,1262,9600,6,TA,Y,Ex,Y,Veenker,1976
2,223500,1786,11250,7,Gd,Y,Ex,Y,CollgCr,2001
3,140000,1717,9550,7,Gd,Y,Gd,Y,Crawfor,1915
4,250000,2198,14260,8,Gd,Y,Ex,Y,NoRidge,2000


---

## 필수 1. 표본 크기와 검정력 비교

### 문제 1-1. 중앙 냉방시설 비교 연구에 필요한 표본 수 설계

#### 문제 설명

중앙 냉방시설이 있는 주택과 없는 주택의 판매가격을 비교하는 연구를 준비하고 있습니다. 사전 조사에서 표준화 효과 크기는 `0.3` 정도로 예상했습니다.

다음 조건에서 그룹당 표본 크기에 따른 검정력을 비교하고, 목표 검정력 0.8을 확보하기 위해 필요한 표본 수를 계산하세요.

- 예상 효과 크기: `0.3`
- 유의수준: `0.05`
- 양측검정
- 비교할 그룹당 표본 수: `30명`, `100명`, `300명`

#### 요구사항

1. `TTestIndPower()` 객체를 생성하세요.
2. 그룹당 표본 수가 30명, 100명, 300명일 때의 검정력을 각각 계산하세요.
3. 표본 크기별 검정력을 소수점 셋째 자리까지 출력하세요.
4. 유의수준 0.05, 목표 검정력 0.8, 효과 크기 0.3일 때 그룹당 필요한 표본 수를 계산하세요.
5. 계산된 표본 수는 소수점 이하를 올림하여 정수로 출력하세요.
6. 세 표본 크기 중 목표 검정력 0.8을 충족하는 경우를 확인하세요.

#### 해석 질문

**Q1.** 효과 크기와 유의수준이 같을 때 표본 크기가 커지면 검정력은 어떻게 변하나요?  
**Q2.** 검정력이 낮으면 실제 차이가 존재할 때 어떤 오류의 위험이 커지나요?  
**Q3.** 필요한 표본 수를 소수점 이하 올림으로 처리하는 이유는 무엇인가요?  
**Q4.** 표본 수를 무조건 크게 설정하는 것이 항상 최선인가요?

#### 제출 결과

- 표본 크기별 검정력
- 목표 검정력에 필요한 그룹당 표본 수
- 목표 검정력 충족 여부
- Q1~Q4 답변

In [2]:
# 필수 1 코드를 작성하세요.


# 1. `TTestIndPower()` 객체를 생성하세요.
# 2. 그룹당 표본 수가 30명, 100명, 300명일 때의 검정력을 각각 계산하세요.
# 3. 표본 크기별 검정력을 소수점 셋째 자리까지 출력하세요.
# 4. 유의수준 0.05, 목표 검정력 0.8, 효과 크기 0.3일 때 그룹당 필요한 표본 수를 계산하세요.
# 5. 계산된 표본 수는 소수점 이하를 올림하여 정수로 출력하세요.
# 6. 세 표본 크기 중 목표 검정력 0.8을 충족하는 경우를 확인하세요.

import math
from statsmodels.stats.power import TTestIndPower

# 1 ~ 3
analysis = TTestIndPower()

power_30 = analysis.power(effect_size=0.3, nobs1 = 30, alpha=0.05)
power_100 = analysis.power(effect_size=0.3, nobs1=100, alpha= 0.05)
power_300 = analysis.power(effect_size=0.3, nobs1=300, alpha=0.05)

print(f'표본크기_30 검정력 : {power_30:.3f}')
print(f'표본크기_100 검정력 : {power_100:.3f}')
print(f'표본크기_300 검정력 : {power_300:.3f}')

# 4
required_n = analysis.solve_power(effect_size=0.3, alpha=0.05, power=0.8)
required_n = math.ceil(required_n)

print(f'그룹당 필요한 표본 수 : {required_n} 명')


표본크기_30 검정력 : 0.208
표본크기_100 검정력 : 0.560
표본크기_300 검정력 : 0.956
그룹당 필요한 표본 수 : 176 명


### 필수 1 답변 작성란

- **Q1.** 효과 크기와 유의수준이 같을 때 표본 크기가 커지면 검정력은 어떻게 변하나요?  

    - 표본 크기가 커지면 일반적으로 검정력이 높아진다.

    - 표본이 많아질수록 실제 차이를 우연한 변동과 구분하기 쉬워짐

<br>

- **Q2.** 검정력이 낮으면 실제 차이가 존재할 때 어떤 오류의 위험이 커지나요?  

    - 실제 차이가 있는데도 귀무가설을 기각하지 못하는 제2종 오류의 위험이 커짐

    - 예를들어 검정력이 0.8 이라면 가정한 실제 차이를 놓칠 확률인 `β`는 `1 - 0.8 = 0.2`이다.
    
<br>

- **Q3.** 필요한 표본 수를 소수점 이하 올림으로 처리하는 이유는 무엇인가요?  

    -  사람이나 주택을 소수 단위로 모집할 수 없고

    - 내림하면 계산된 목표 검정력을 충족하지 못할 수 있다

<br>

- **Q4.** 표본 수를 무조건 크게 설정하는 것이 항상 최선인가요?

    - 아닙니다. 표본 확보에는 비용과 시간이 소요가 되기 때문.

---

## 필수 2. 통계적 유의성과 실질적 유의성 종합 판단

### 문제 2-1. 중앙 냉방시설 유무에 따른 판매가격 차이 분석

#### 문제 설명

한 부동산 회사는 중앙 냉방시설이 있는 주택과 없는 주택의 평균 판매가격 차이를 분석하려고 합니다. 회사는 두 집단의 평균 판매가격 차이가 **100,000달러 이상**이어야 냉방시설 설치 지원 사업을 검토할 실무적 가치가 있다고 정했습니다.

#### 요구사항

1. `CentralAir == "Y"`인 주택의 `SalePrice`를 `air_yes`에 저장하세요.
2. `CentralAir == "N"`인 주택의 `SalePrice`를 `air_no`에 저장하세요.
3. 두 집단의 표본 수와 평균 판매가격을 출력하세요.
4. `stats.ttest_ind()`로 두 집단 평균 차이에 대한 양측검정을 수행하세요.
5. 제공된 공식에 따라 Cohen’s d 계산 함수를 작성하고 효과 크기를 구하세요.
6. 평균 차이, p-value, Cohen’s d를 출력하세요.
7. 다음 기준으로 결과를 판단하세요.
   - `p-value ≤ 0.05`: 통계적으로 유의함
   - `|Cohen’s d| ≥ 0.8`: 큰 효과
   - `|평균 차이| ≥ 100000`: 회사 기준에서 실질적으로 유의함
8. 세 결과를 종합하여 사업 검토 여부에 대한 근거를 작성하세요.

#### Cohen’s d 공식

두 집단의 평균 차이를 합동 표준편차로 나누어 계산합니다.

$$
d = \frac{\bar{x}_1 - \bar{x}_2}{s_p}
$$

$$
s_p = \sqrt{\frac{(n_1-1)s_1^2 + (n_2-1)s_2^2}{n_1+n_2-2}}
$$

- $\bar{x}_1$, $\bar{x}_2$: 각 집단의 평균
- $s_1$, $s_2$: 각 집단의 표본 표준편차
- $n_1$, $n_2$: 각 집단의 표본 수
- $s_p$: 합동 표준편차

#### 해석 질문

**Q1.** p-value와 Cohen’s d는 각각 어떤 정보를 제공하나요?  
**Q2.** 두 집단의 판매가격 차이는 통계적으로 유의한가요?  
**Q3.** Cohen’s d를 기준으로 효과 크기는 어느 정도인가요?  
**Q4.** 회사가 정한 100,000달러 기준에서 실질적으로 유의한가요?  
**Q5.** 통계적으로 유의하고 효과 크기가 크더라도 회사의 실무 기준을 충족하지 못할 수 있나요?

#### 제출 결과

- 집단별 표본 수와 평균
- 독립표본 t검정 결과
- 평균 차이와 Cohen’s d
- 통계적·효과크기·회사 기준 판단
- 최종 의사결정 근거
- Q1~Q5 답변

In [3]:
# 필수 2 코드를 작성하세요.


air_yes = df[df['CentralAir'] == 'Y']['SalePrice']
air_no = df[df['CentralAir'] == 'N']['SalePrice']

print(f"air_yes 표본 수 : {len(air_yes)} | 평균가격 : {air_yes.mean():.3f}")
print(f"air_no 표본 수 : {len(air_no)} | 평균가격 : {air_no.mean():.3f}")

# 독립표본 t-test
t_stats, pvalue = stats.ttest_ind(air_yes, air_no)

print(f'통계량 : {t_stats:.3f}, p-value : {pvalue:.3f}')


# Cohen's d
mean_yes = air_yes.mean()
mean_no = air_no.mean()

std_yes = air_yes.std()
std_no = air_no.std()

## Pooled Standard Deviation
n_yes = len(air_yes)
n_no = len(air_no)

pooled_std = np.sqrt((n_yes - 1) * std_yes ** 2 + (n_no - 1) * std_no ** 2 / (n_yes + n_no - 2))

## Cohen's d
cohens_d = (mean_no - mean_yes) / pooled_std

## 독립표본 t 검정
t_stats, pvalue = stats.ttest_ind(air_yes, air_no, equal_var=True)


print()
print(f"평균 차이 : {mean_no - mean_yes}")
print(f"Cohen's d: {cohens_d:.3f}")
print(f"p-value : {pvalue:.3f}")

if pvalue <= 0.05:
    print('통계적으로 유의함')
else:
    print('통계적으로 유의하지 않음')

if cohens_d >= 0.8:
    print('cohens_d 큰 효과')
else:
    print('cohens_d 미미한 효과')

air_yes 표본 수 : 1365 | 평균가격 : 186186.710
air_no 표본 수 : 95 | 평균가격 : 105264.074
통계량 : 9.915, p-value : 0.000

평균 차이 : -80922.63620589937
Cohen's d: -0.028
p-value : 0.000
통계적으로 유의함
cohens_d 미미한 효과


### 필수 2 답변 작성란

- **Q1.** p-value와 Cohen’s d는 각각 어떤 정보를 제공하나요?  

    - p-value는 관측된 차이가 통계적으로 유의한지 판단하는 데 사용

    - Cohen's d는 두 집단 차이의 크기를 표준화하여 보여줌

<br>

- **Q2.** 두 집단의 판매가격 차이는 통계적으로 유의한가요?  

    - 유의함. p-value가 0.05보다 작음

    - 귀무가설: 냉방시설 유무는 주택 가격 평균에 차이를 보이지 않는다 -> 기각

<br>

- **Q3.** Cohen’s d를 기준으로 효과 크기는 어느 정도인가요?  

    - Cohen's의 절대값이 1.05고 규정한 0.8보다 크기 때문에 큰 효과로 해석할 수 있다.

<br>

- **Q4.** 회사가 정한 100,000달러 기준에서 실질적으로 유의한가요?  
    
    - 유의하지않음, 80.900달러 정도였기 떄문에 회사의 기대 미충족 

<br>

- **Q5.** 통계적으로 유의하고 효과 크기가 크더라도 회사의 실무 기준을 충족하지 못할 수 있나요?

    - 그렇다. 

    - 통계적 유의성은 차이가 있다는 근거가 있는가?

    - 효과 크기는 자료의 흩어짐에 비해 차이가 큰가?
    
    - 회사 입장은 업무에서 요구하는 최소 금액 차이를 넘는가?

---

## 과제. 주방 품질에 따른 판매가격 차이 분석

### 문제 3-1. 주방 품질이 `Gd`인 집단과 `TA`인 집단 비교

#### 문제 설명

부동산 중개회사는 주방 품질이 `Gd`(Good)인 주택과 `TA`(Typical/Average)인 주택의 평균 판매가격을 비교하려고 합니다. 회사는 평균 판매가격 차이가 **50,000달러 이상**이면 마케팅에서 주방 품질의 차이를 강조할 실무적 가치가 있다고 판단합니다.

> 필수 2에서 학습한 검정과 효과 크기 계산 절차를 새로운 두 집단에 적용하는 과제입니다.

#### 요구사항

1. 주방 품질이 `Gd`인 집단과 `TA`인 집단의 `SalePrice`를 각각 준비하세요.
2. 두 집단의 표본 수와 평균 판매가격을 출력하세요.
3. 두 집단의 평균 차이에 대한 양측 독립표본 t검정을 수행하세요.
4. Cohen’s d를 계산하세요.
5. 평균 차이, p-value, Cohen’s d를 출력하세요.
6. 다음 기준에 따라 각각 판단하세요.
   - 통계적 유의성: `p-value ≤ 0.05`
   - 큰 효과: `|Cohen’s d| ≥ 0.8`
   - 실무적 유의성: `|평균 차이| ≥ 50000`
7. 세 가지 판단을 종합하여 주방 품질을 마케팅에서 강조할 근거가 있는지 결론을 작성하세요.

#### 해석 질문

**Q1.** 두 집단의 판매가격 차이는 통계적으로 유의한가요?  
**Q2.** Cohen’s d를 기준으로 효과 크기는 어느 정도인가요?  
**Q3.** 회사가 정한 50,000달러 기준을 충족하나요?  
**Q4.** 최종적으로 주방 품질 차이를 마케팅에서 강조할 근거가 있다고 볼 수 있나요?

#### 제출 결과

- 집단별 표본 수와 평균
- t통계량과 p-value
- 평균 차이와 Cohen’s d
- 통계적·실질적 유의성 판단
- 최종 결론
- Q1~Q4 답변

In [4]:
# 과제 코드를 작성하세요.

# 1. 주방 품질이 `Gd`인 집단과 `TA`인 집단의 `SalePrice`를 각각 준비하세요.
# 2. 두 집단의 표본 수와 평균 판매가격을 출력하세요.
# 3. 두 집단의 평균 차이에 대한 양측 독립표본 t검정을 수행하세요.
# 4. Cohen’s d를 계산하세요.
# 5. 평균 차이, p-value, Cohen’s d를 출력하세요.

kitchen_gd = df[df['KitchenQual'] == 'Gd']['SalePrice']
kitchen_ta = df[df['KitchenQual'] == 'TA']['SalePrice']

n_gd = len(kitchen_gd)
n_ta = len(kitchen_ta)

mean_gd = kitchen_gd.mean()
mean_ta = kitchen_ta.mean()

std_gd = kitchen_gd.std()
std_ta = kitchen_ta.std()

print(f'주방품질 Gd 표본 수 : {n_gd} | 평균 판매가격 : {mean_gd:.3f}')
print(f'주방품질 TA 표본 수 : {n_ta} | 평균 판매가격 : {mean_ta:.3f}')


## Pooled Standard Deviation
pooled_std = np.sqrt((n_gd - 1) * std_gd ** 2 + (n_ta - 1) * std_ta ** 2 / (n_gd + n_ta - 2))

## Cohen's d
cohens_d = (mean_ta - mean_gd) / pooled_std


# 독립표본 t-test
t_stats, pvalue = stats.ttest_ind(kitchen_gd, kitchen_ta)

print()
print(f'평균 차이 : {mean_ta - mean_gd:.3f}')
print(f"Cohen's d: {cohens_d:.3f}")
print(f"p-value : {pvalue:.3f}")
print(f't 통계량 : {t_stats:.3f}')

# 유의수준
if pvalue <= 0.05:
    print('통계적으로 유의함')
else:
    print('통계적으로 유의하지 않음')

# cohens_d
if cohens_d >= 0.8:
    print('cohens_d 큰 효과')
else:
    print('cohens_d 미미한 효과')

주방품질 Gd 표본 수 : 586 | 평균 판매가격 : 212116.024
주방품질 TA 표본 수 : 735 | 평균 판매가격 : 139962.512

평균 차이 : -72153.512
Cohen's d: -0.047
p-value : 0.000
t 통계량 : 25.263
통계적으로 유의함
cohens_d 미미한 효과


### 과제 답변 작성란

- **Q1.** 두 집단의 판매가격 차이는 통계적으로 유의한가요?

    - 두 집단의 평균 차이 비교검정인 독립표본 t-test 를 진행하여 

    - 유의수준인 p-value는 `0.000`으로 기준 유의수준인 0.005보다 낮기 때문에

    - 두 집단의 판매가격 차이는 통계적으로 유의하다고 판단이 가능하다.

<br>    

- **Q2.** Cohen’s d를 기준으로 효과 크기는 어느 정도인가요?  

    - Cohen’s d는 `-0.047`으로 나타났다.

    - `Cohen’s d = |-0.47| = 0.47`로 효과크기는 중간효과가 나타났다고 볼 수 있다.

<br>    

- **Q3.** 회사가 정한 50,000달러 기준을 충족하나요?  

    - 충촉한다. 평균 차이는 `-72153.512`로 나타났기 때문이다.
 
<br>   

- **Q4.** 최종적으로 주방 품질 차이를 마케팅에서 강조할 근거가 있다고 볼 수 있나요?

    - 두 집단의 평균차이는 통계적으로 유의하기 때문에 주방 품질 차이는 존재한다고 볼 수 있다.

    - 따라서 주방 품질 차이를 마케팅으로 강조할 근거가 있다고 판단된다.

---

## 실습 마무리

1. 검정력은 무엇이며 제2종 오류와 어떤 관계가 있나요?
2. 같은 효과 크기와 유의수준에서 표본 크기가 커지면 검정력은 어떻게 변하나요?
3. 목표 검정력이 높거나 발견하려는 효과가 작을수록 필요한 표본 수는 어떻게 변하나요?
4. p-value와 Cohen’s d를 함께 확인해야 하는 이유는 무엇인가요?
5. 통계적으로 유의한 결과가 반드시 실무적으로 중요한 결과를 의미하나요?